### PR TITLE
Jsx3 andrii

### PR DESCRIPTION
**Pull request by andrii-antiukhov**
In this files i made changes, but not all was refactored to JSX3 (need to do this)

```
client/shared/Fi.re
client/shared/React.re
client/shared/ReactOld.re
```

All components in folder `client/src/components/` was refactored to JSX3. But i did not run compiler, so maybe there are some errors.
@jchavarri Not sure about changing `React.re` to `ReactOld.re` I did this because last time me and Feihong found bug, so i created other `React.re` But i do not know do we need this for JSX3 Sorry. If not need thiod changes, i will remove `ReactOld` from all files in this PR.

### CODE REVIEW

File: client/shared/ButterToast.re [hunk 1]
```diff
@@ -69,7 +69,7 @@ module ToastOption = {
 
   [@bs.deriving abstract]
   type t = {
-    content: contentArgs => React.reactElement,
+    content: contentArgs => ReactOld.reactElement,
```
Comment (jchavarri) [inline client/shared/ButterToast.re:72]: Using `ReactOld` temporarily is ok while you work on this, but ultimately it's not needed and `ReactOld` should be removed before we merge the jsx3 migration.

File: client/shared/Fi.re [hunk 2]
```diff
@@ -1,248 +1,94 @@
-/* react-icons: feather icon */
-module IconContext = {
```
Comment (jchavarri) [inline client/shared/Fi.re:2]: This was not used?

File: client/src/Router.re [hunk 3]
```diff
@@ -51,7 +51,7 @@ module Unload = {
         );
         None;
       });
-      React.null;
+      ReactOld.null;
```
Comment (jchavarri) [inline client/src/Router.re:54]: I don't understand this change :/ why not using `React.null`? One less thing to change later on.

File: client/src/Router.rei [hunk 4]
```diff
@@ -23,7 +23,8 @@ module Unload: {
    */
   module Provider: {
     let make:
-      React.childless => React.component(unit, React.noRetainedProps, unit);
+      ReactOld.childless =>
+      ReactOld.component(unit, ReactOld.noRetainedProps, unit);
```
Comment (jchavarri) [inline client/src/Router.rei:27]: We should update these signatures to jsx3 as well, e.g.

```reason
[@react.component]
let make: unit => React.element;
```

File: client/src/Router.rei [hunk 5]
```diff
@@ -41,7 +42,11 @@ module Link: {
       ~className: string=?,
       ~popup: bool=?,
       ~role: string=?,
-      array(React.reactElement)
+      array(ReactOld.reactElement)
```
Comment (jchavarri) [inline client/src/Router.rei:45]: We should update this to jsx3 as well:
```reason
[@react.component]
  let make:
    (
       ~route: Route.t,
      ~id: string=?,
      ~title: string=?,
      ~className: string=?,
      ~popup: bool=?,
      ~role: string=?,
      ~children: React.element,
      unit
    ) =>
    React.element;
```

File: client/src/Router.rei [hunk 6]
```diff
@@ -41,7 +42,11 @@ module Link: {
...
     ) =>
-    React.component(React.stateless, React.noRetainedProps, React.actionless);
+    ReactOld.component(
+      ReactOld.stateless,
+      ReactOld.noRetainedProps,
+      ReactOld.actionless,
+    );
```
Comment (jchavarri) [inline client/src/Router.rei:51]: This should be just `React.element`

File: client/src/components/Notify.re [hunk 7]
```diff
@@ -1,6 +1,7 @@
-Modules.require("./Notify.css");
+Modules.import("./Notify.css");
```
Comment (jchavarri) [inline client/src/components/Notify.re:1]: Please undo all the changes from `require` to `import`. They will break the app. We want `require`.

File: client/src/components/Notify.re [hunk 8]
```diff
@@ -1,6 +1,7 @@
...
 open Utils;
 
+[@react.component]
 let base = (~className, ~sticky=false, message) =>
```
Comment (jchavarri) [inline client/src/components/Notify.re:5]: `message` should be a labelled arg. Probably it was meant to be `children` 🤔

File: client/src/components/UI_DateTime.re [hunk 9]
```diff
@@ -19,22 +19,12 @@ module Transformer: {
...
-    | None => ReasonReact.null
-    | Some(date) =>
-      <time dateTime=date ?className> {date->transformer->str} </time>
-    };
-  },
+[@react.component]
+let make = (~date: Js.Json.t, ~transformer=Transformer.relative, ~className=?) => {
+  let date = date |> Js.Json.decodeString;
+  switch (date) {
+  | None => ReasonReact.null
```
Comment (jchavarri) [inline client/src/components/UI_DateTime.re:26]: We can use `React.null` now.

File: client/src/components/UI_LoadingWrapper.re [hunk 10]
```diff
@@ -8,39 +8,44 @@ type action =
...
+  let (state, dispatch) =
+    React.useReducer(
+      reducer,
+      {loading, displayLoading: loading, timeoutId: ref(None)},
+    );
+
+  let {loading, displayLoading, timeoutId} = state;
+
+  React.useEffect(() => {
+    clearTimeout(timeoutId^);
```
Comment (jchavarri) [inline client/src/components/UI_LoadingWrapper.re:33]: If you look carefully at the previous code, this call to `clearTimeout` was done when unmounting. To keep the same behavior, it should be moved to the returned value of the effect, e.g.

```
 React.useEffect(() => {
    Some(() => clearTimeout(timeoutId^));
}
```

File: client/src/components/UI_LoadingWrapper.re [hunk 11]
```diff
@@ -49,15 +54,13 @@ let make = (~delayMs=1000, ~loading: bool, children) => {
...
-    switch (action) {
-    | ChangeLoadingDisplay(loading) =>
-      ReasonReact.Update({...state, displayLoading: loading})
-    | ChangeLoading(loading) => ReasonReact.Update({...state, loading})
-    },
-  render: ({state}) => children(state.displayLoading),
+    [|loading|],
+  );
+
+  displayLoading ? children : ReactOld.null;
```
Comment (jchavarri) [inline client/src/components/UI_LoadingWrapper.re:65]: This is not the same behavior as before. It should be something like `children(displayLoading)`.

File: client/src/components/UI_UserInfo.re [hunk 12]
```diff
@@ -1,46 +1,42 @@
...
-                         alt={user##username ++ " avatar"}
-                       />
-                   )}
-                </dl>
-              )
+[@react.component]
+let make = (~userId) => {
+  let query = UserInfoGql.make(~userId, ());
+  <UserInfoComponent variables=query##variables>
+    ...{({result}) =>
```
Comment (jchavarri) [inline client/src/components/UI_UserInfo.re:8]: I don't think spread operator `...` is needed with jsx3. I am pretty sure this won't compile.

**Review by jchavarri — COMMENTED** [review]
@andrii-antiukhov this is looking good. please take a look at comment, I will merge after they are tackled, thanks!

File: client/shared/ButterToast.re (hunk 1 above)
Comment (andrii-antiukhov) [reply client/shared/ButterToast.re]: Removed `ReacOld`

File: client/shared/Fi.re (hunk 2 above)
Comment (andrii-antiukhov) [reply client/shared/Fi.re:2]: Looks like this icon does not used in code.

File: client/src/Router.re (hunk 3 above)
Comment (andrii-antiukhov) [reply client/src/Router.re]: Remove `ReactOld`

File: client/src/components/Notify.re (hunk 7 above)
Comment (andrii-antiukhov) [reply client/src/components/Notify.re]: Returned back.

File: client/src/Router.rei (hunk 4 above)
Comment (andrii-antiukhov) [reply client/src/Router.rei]: Changed.

File: client/src/Router.rei (hunk 5 above)
Comment (andrii-antiukhov) [reply client/src/Router.rei]: Changed.

File: client/src/Router.rei (hunk 6 above)
Comment (andrii-antiukhov) [reply client/src/Router.rei]: Changed.

File: client/src/components/Notify.re (hunk 8 above)
Comment (andrii-antiukhov) [reply client/src/components/Notify.re]: I changed it like labeled. But according code this variable must be string type, not children. Maybe.

File: client/src/components/UI_DateTime.re (hunk 9 above)
Comment (andrii-antiukhov) [reply client/src/components/UI_DateTime.re]: Replaced `ReasonReact.null` to `React.null`

File: client/src/components/UI_LoadingWrapper.re (hunk 10 above)
Comment (andrii-antiukhov) [reply client/src/components/UI_LoadingWrapper.re]: Thanks! Changed to return some value.

File: client/src/components/UI_UserInfo.re (hunk 12 above)
Comment (andrii-antiukhov) [reply client/src/components/UI_UserInfo.re]: Removed

File: client/src/components/UI_LoadingWrapper.re (hunk 11 above)
Comment (andrii-antiukhov) [reply client/src/components/UI_LoadingWrapper.re]: Sorry, i am not sure `children` can be function. But maybe. Thanks!

**Review by jchavarri — APPROVED** [review]
Looks good! Please take a look at these minor things and I think we can merge it ,thanks!

File: client/src/components/UI_DateTime.re [hunk 13]
```diff
@@ -19,22 +19,12 @@ module Transformer: {
...
-  render: _self => {
-    let date = date |> Js.Json.decodeString;
-    switch (date) {
-    | None => ReasonReact.null
-    | Some(date) =>
-      <time dateTime=date ?className> {date->transformer->str} </time>
-    };
-  },
+[@react.component]
+let make = (~date: Js.Json.t, ~transformer=Transformer.relative, ~className=?) => {
```
[the remark this answers is not in the document]
Comment (maxlibin) [reply client/src/components/UI_DateTime.re:23]: This component is already updated in jsx3 branch: https://github.com/Sketch-sh/sketch-sh/pull/321
Screenshot 2022-05-06 at 1 51 57 PM

[the remark this answers is not in the document]
Comment (andrii-antiukhov) [reply client/src/components/UI_DateTime.re:23]: @jchavarri @maxlibin Here i added this `[@react.component]` I think it need this code. If no need i will remove.

File: client/src/components/UI_SketchOwnerInfo.re [hunk 14]
```diff
@@ -1,37 +1,35 @@
...
-               alt={owner##username ++ " avatar"}
-             />
-         )}
-        <span className="UI_SketchOwnerInfo__username">
-          {owner##username->str}
-        </span>
-      </Router.Link>
-      {noteLastEdited
+[@react.component]
+let make = (~owner, ~noteLastEdited=?, ~className=?) => {
```
[the remark this answers is not in the document]
Comment (andrii-antiukhov) [reply client/src/components/UI_SketchOwnerInfo.re]: Changed.

File: client/src/components/UI_FullpageLoading.re [hunk 15]
```diff
@@ -1,4 +1,4 @@
-Modules.import("./UI_FullpageLoading.css");
+Modules.require("./UI_FullpageLoading.css");
```
[the remark this answers is not in the document]
Comment (andrii-antiukhov) [reply client/src/components/UI_FullpageLoading.re]: Changed.

**Comment by andrii-antiukhov** [comment]
Hi @jchavarri Today i successfully up sketch on my local Mac and i did some refactoring. So not so bad was today. I refactored `Embed.re` and `ReLoadable.re` and fixed some other issues. Please, review this PR and you can merge it to branch `jsx3` if need.
For me and @maxlibin left to refactor 2 files
```
 src_editor/Editor_Blocks.re
 src_editor/Editor_Links.re
```
But maybe we will do this next time, because this files not so simple to refactor. Thanks!

**Comment by andrii-antiukhov** [comment]
@jchavarri Sorry i can not finish properly file `src_editor/Editor_Blocks.re` too complicated for me. I done main part but i do not know how to refactor some old code. So you can look and maybe can fix rest part of code. Sorry.
File `src_editor/Editor_Links.re` was refactored full. Sorry, i do not have types because can not up compiler, difficult to understand logic. So i hope i help you how i can. I tried do everything in best way, but not enough knowledge.
So me and Max waiting for your instructions for the next opensource day. If it is still make sense for us. Thanks!

**Pull request closed without merging**